pinot-query-planner/src/main/java/org/apache/pinot/query/routing/WorkerMetadata.java

```diff
@@ -0,0 +1,107 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.query.routing;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.core.type.TypeReference;
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import org.apache.pinot.spi.utils.JsonUtils;
+
+
+/**
+ * The {@code WorkerMetadata} info contains the information for executing a particular worker on a particular stage.
+ *
+ * <p>It contains information about:</p>
+ * <ul>
+ *   <li>the tables it is suppose to scan (including real-time and offline portions separately)</li>
+ *   <li>the underlying segments a stage requires to execute upon if it is a leaf stage worker</li>
+ *   <li>the partition mechanism of the data being execute on this worker</li>
+ * </ul>
+ *
+ * <p>WorkerMetadata should allow {@link org.apache.pinot.query.planner.stage.MailboxSendNode} and
+ * {@link org.apache.pinot.query.planner.stage.MailboxReceiveNode} to derive the appropriate set of mailboxes to be
+ * created when executing the worker.</p>
+ */
+public class WorkerMetadata {
+  private final VirtualServerAddress _virtualServerAddress;
+  private final Map<String, String> _customProperties;
+
+  private WorkerMetadata(VirtualServerAddress virtualServerAddress, Map<String, String> customProperties) {
+    _virtualServerAddress = virtualServerAddress;
+    _customProperties = customProperties;
+  }
+
+  public VirtualServerAddress getVirtualServerAddress() {
+    return _virtualServerAddress;
+  }
+
+  public Map<String, String> getCustomProperties() {
+    return _customProperties;
+  }
+
+  public static class Builder {
+    public static final String TABLE_SEGMENTS_MAP_KEY = "tableSegmentsMap";
+    private VirtualServerAddress _virtualServerAddress;
+    private Map<String, String> _customProperties;
+
+    public Builder() {
+      _customProperties = new HashMap<>();
+    }
+
+    public Builder setVirtualServerAddress(VirtualServerAddress virtualServerAddress) {
+      _virtualServerAddress = virtualServerAddress;
+      return this;
+    }
+
+    public Builder addTableSegmentsMap(Map<String, List<String>> tableSegmentsMap) {
+      try {
+        String tableSegmentsMapStr = JsonUtils.objectToString(tableSegmentsMap);
+        _customProperties.put(TABLE_SEGMENTS_MAP_KEY, tableSegmentsMapStr);
+      } catch (JsonProcessingException e) {
+        throw new RuntimeException("Unable to serialize table segments map", e);
+      }
+      return this;
+    }
+
+    public WorkerMetadata build() {
+      return new WorkerMetadata(_virtualServerAddress, _customProperties);
+    }
+
+    public void putAllCustomProperties(Map<String, String> customPropertyMap) {
+      _customProperties.putAll(customPropertyMap);
+    }
+  }
+
+  public static Map<String, List<String>> getTableSegmentsMap(WorkerMetadata workerMetadata) {
```

Comment:
we can refactor this into a util going forward if we decided upon a common string ser/de mechanism for all custom properties.